File: greedy/56.merge-intervals.py
#
# @lc app=leetcode id=56 lang=python3
#
# [56] Merge Intervals
#

# @lc code=start
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals = sorted(intervals)
        res = []
        prev_start = intervals[0][0]
        prev_end = intervals[0][1]

        for start, end in intervals[1:]:
            if start <= prev_end:
                prev_end = max(prev_end, end)
            else:
                res.append([prev_start, prev_end])
                prev_start = start
                prev_end = end

        res.append([prev_start, prev_end])
        return res


# @lc code=end

File: Interval/56.merge-intervals.java
/*
 * @lc app=leetcode id=56 lang=java
 *
 * [56] Merge Intervals
 */
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// @lc code=start
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        List<int[]> res = new ArrayList<>();
        int left = intervals[0][0];
        int right = intervals[0][1];
        for (int i = 1; i < intervals.length; i++) {
            if (right >= intervals[i][0]) {
                right = Math.max(intervals[i][1], right);
            } else {
                res.add(new int[] {left, right});
                left = intervals[i][0];
                right = intervals[i][1];
            }
        }
        res.add(new int[] {left, right});
        return res.toArray(new int[res.size()][]);
    }
}
// @lc code=end
